### ISSUE TITLE
Tasks are not impodent

### ISSUE THREAD

**Issue by WilliamDEdwards** [sensu/sensu-go-ansible#198]
At least the following tasks are not impodent:

- bonsai_asset
- filter
- pipe_handlers
- handler_set
- check
- entity

This results in tasks reporting the `changed` state when not changed, while they should report `ok` state.

**Comment by tadeboro** [comment]
Hi, @WilliamDEdwards. Can you please provide a version number of the Sensu Go backend that you are using, version of the Sensu Go Ansible Collection that you have installed, and a sample playbook that triggers the error?

We do have quite a few tests for idempotency ([bonsai asset](https://github.com/sensu/sensu-go-ansible/blob/7309c1259b7d5b2aff15bdbf52a84d5c5ee9d625/tests/integration/molecule/action_bonsai_asset/converge.yml#L24), [filter](https://github.com/sensu/sensu-go-ansible/blob/7309c1259b7d5b2aff15bdbf52a84d5c5ee9d625/tests/integration/molecule/module_filter/converge.yml#L68), [pipe_handler](https://github.com/sensu/sensu-go-ansible/blob/7309c1259b7d5b2aff15bdbf52a84d5c5ee9d625/tests/integration/molecule/module_pipe_handler/converge.yml#L65), [handler_set](https://github.com/sensu/sensu-go-ansible/blob/7309c1259b7d5b2aff15bdbf52a84d5c5ee9d625/tests/integration/molecule/module_handler_set/converge.yml#L24), [check](https://github.com/sensu/sensu-go-ansible/blob/7309c1259b7d5b2aff15bdbf52a84d5c5ee9d625/tests/integration/molecule/module_check/converge.yml#L91), [entity](https://github.com/sensu/sensu-go-ansible/blob/7309c1259b7d5b2aff15bdbf52a84d5c5ee9d625/tests/integration/molecule/module_entity/converge.yml#L81)) and currently, they are all green.

We had some troubles with idempotency in the past because Sensu Go 5.19.0 changed its API a bit, but this should be all resolved since version 1.4.0 of collection, released more than a month ago.

**Comment by WilliamDEdwards** [comment]
You're right. I was using version 1.3.1.

Thanks.

**Comment by WilliamDEdwards** [comment]
Actually, this does not seem to be completely fixed; creating checks without runtime assets (or an empty list!) returns 'changed'.

**Comment by tadeboro** [comment]
Can you please provide a sample Ansible task that demonstrates the failure? I cannot replicate the issue with the `sensu.sensu_go.check` invocations I have at hand.

**Comment by WilliamDEdwards** [comment]
1.

```
   - name: Create proxy checks
      check:
        name: "{{ item.name }}"
        command: "{{ item.command }}"
        subscriptions: "{{ item.subscribers }}"
        interval: "{{ item.interval|default(180) }}"
        handlers: "{{ item.handlers|default(['email', 'pushover']) }}"
        round_robin: "{{ item.round_robin|default(true) }}"
        low_flap_threshold: 20
        high_flap_threshold: 60
        publish: true
        proxy_requests:
          entity_attributes:
            - "entity.entity_class == 'proxy'"
            - "{% for entity in item.entities %}entity.labels.proxy_type == '{{ entity }}'{% if not loop.last %} || {% endif %}{% endfor %}"
        auth: "{{ sensu_auth_line }}"
      with_items: "{{ sensu_proxy_checks }}"
```

This reports 'changed'.

```
    - name: Create checks
      check:
        name: "{{ item.name }}"
        command: "{{ item.command }}"
        subscriptions: "{{ item.subscribers|default([]) }}"
        interval: "{{ item.interval|default(180) }}"
        handlers: "{{ item.handlers|default(['email', 'pushover']) }}"
        low_flap_threshold: 20
        high_flap_threshold: 60
        publish: true
        runtime_assets: "{{ item.runtime_assets|default([]) }}"
        auth: "{{ sensu_auth_line }}"
      with_items: "{{ sensu_checks }}"
```

This reports 'changed' for items where `runtime_assets` is `[]`.

**Comment by tadeboro** [comment]
I can confirm the problems with the check module. We are too strict when it comes to some complex parameter comparisons (proxy_requests, handlers, runtime_assets).

**Comment by tadeboro** [comment]
@WilliamDEdwards I created a fix for the issues with the check module. You can test it if you wish by cloning the `check-idempotency` branch.

**Comment by WilliamDEdwards** [comment]
Hi @tadeboro, thanks for the quick fix. Really cool.

**Comment by tadeboro** [comment]
Just a heads up: this fix is now part of the 1.4.2 release.
